snafu/upgrade_openshift_wrapper/trigger_upgrade.py

```diff
@@ -0,0 +1,166 @@
+#!/usr/bin/env python
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+
+import logging
+import time
+import datetime
+import subprocess
+from kubernetes import client, config
+from openshift.dynamic import DynamicClient
+
+logger = logging.getLogger("snafu")
+time_format = "%Y-%m-%dT%H:%M:%S%z"
+
+class Trigger_upgrade():
+    def __init__(self, args):
+        self.uuid = args.uuid
+        self.user = args.user
+        self.version = args.version
+        self.major = args.major
+        self.cluster_name = args.cluster_name
+        self.incluster = args.incluster
+        self.poll_interval = args.poll_interval
+        self.timeout = args.timeout
+        self.kubeconfig = args.kubeconfig
+        self.toimage = args.toimage
+        if self.incluster == "true":
+            config.load_incluster_config()
+            k8s_config = client.Configuration()
+            k8s_client = client.api_client.ApiClient(configuration=k8s_config)
+        elif self.kubeconfig:
+            k8s_client = config.new_client_from_config(self.kubeconfig)
+        else:
+            k8s_client = config.new_client_from_config()
+
+        self.dyn_client = DynamicClient(k8s_client)
+
+    def _json_payload(self, data):
+        payload = {
+            "uuid": self.uuid,
+            "cluster_name": self.cluster_name,
+            "init_version": self.init_version,
+            "end_version": self.end_version,
+            "timestamp": self.timestamp,
+            "platform": self.platform,
+            "start_time": self.start_time.strftime(time_format),
+        }
+        payload.update(data)
+        return payload
+
+    def _run_upgrade(self):
+        # Get platform
+        infra = self.dyn_client.resources.get(kind='Infrastructure')
+        platform = infra.get().attributes.items[0].spec.platformSpec.type or "Unknown"
+
+        clusterversion = self.dyn_client.resources.get(kind='ClusterVersion')
+
+        ver = ("oc get clusterversions.config.openshift.io | grep version | awk '{print $2}'")
```

Comment:
you can do
```
(Pdb) json.loads(oc.selector(['clusterversion']).object_json())["items"][0]["status"]["desired"]["version"]
'4.5.0-0.nightly-2020-09-28-124031'
```

Reply:
Updated. Using a different method but effectively the same.

Reply:
@mohit-sheth so since this change i get During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/root/plow/workloads/upgrade-perf/upgrade/bin/run_snafu", line 11, in <module>
    load_entry_point('snafu', 'console_scripts', 'run_snafu')()
  File "/tmp/snafu/snafu/run_snafu.py", line 101, in main
    res_beg, res_end, res_suc, res_dup, res_fail, res_retry = streaming_bulk(es,
  File "/tmp/snafu/snafu/utils/py_es_bulk.py", line 166, in streaming_bulk
    for ok, resp_payload in streaming_bulk_generator:
  File "/root/plow/workloads/upgrade-perf/upgrade/lib64/python3.8/site-packages/elasticsearch/helpers/actions.py", line 211, in streaming_bulk
    for bulk_data, bulk_actions in _chunk_actions(
  File "/root/plow/workloads/upgrade-perf/upgrade/lib64/python3.8/site-packages/elasticsearch/helpers/actions.py", line 63, in _chunk_actions
    for action, data in actions:
  File "/tmp/snafu/snafu/utils/py_es_bulk.py", line 114, in actions_tracking_closure
    for cl_action in cl_actions:
  File "/tmp/snafu/snafu/run_snafu.py", line 141, in process_generator
    for action, index in data_object.emit_actions():
  File "/tmp/snafu/snafu/upgrade_openshift_wrapper/trigger_upgrade.py", line 145, in emit_actions
    self._run_upgrade()
  File "/tmp/snafu/snafu/upgrade_openshift_wrapper/trigger_upgrade.py", line 96, in _run_upgrade
    if clusterversion.get().attributes.items[0].status.history[0].state == "Completed" \
  File "/root/plow/workloads/upgrade-perf/upgrade/lib64/python3.8/site-packages/openshift/dynamic/client.py", line 94, in get
    return self.request('get', path, **kwargs)
  File "/root/plow/workloads/upgrade-perf/upgrade/lib64/python3.8/site-packages/openshift/dynamic/client.py", line 44, in inner
    raise api_exception(e)
openshift.dynamic.exceptions.InternalServerError: 500
Reason: Internal Server Error
HTTP response headers: HTTPHeaderDict({'Audit-Id': '856e2af0-7cf2-432e-a0b2-b53975a40f4b', 'Cache-Control': 'no-cache, private', 'Content-Type': 'application/json', 'X-Kubernetes-Pf-Flowschema-Uid': 'e95e802a-3b0d-4252-a464-d0c44f981a93', 'X-Kubernetes-Pf-Prioritylevel-Uid': 'bf0545c4-62db-46a7-8fef-e2c89d9eeee2', 'Date': 'Fri, 09 Oct 2020 13:18:01 GMT', 'Content-Length': '119'})
HTTP response body: b'{"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"etcdserver: leader changed","code":500}\n'
Original traceback: 
  File "/root/plow/workloads/upgrade-perf/upgrade/lib64/python3.8/site-packages/openshift/dynamic/client.py", line 42, in inner
    resp = func(self, *args, **kwargs)

  File "/root/plow/workloads/upgrade-perf/upgrade/lib64/python3.8/site-packages/openshift/dynamic/client.py", line 235, in request
    return self.client.call_api(

  File "/root/plow/workloads/upgrade-perf/upgrade/lib64/python3.8/site-packages/kubernetes/client/api_client.py", line 340, in call_api
    return self.__call_api(resource_path, method,

  File "/root/plow/workloads/upgrade-perf/upgrade/lib64/python3.8/site-packages/kubernetes/client/api_client.py", line 172, in __call_api
    response_data = self.request(

  File "/root/plow/workloads/upgrade-perf/upgrade/lib64/python3.8/site-packages/kubernetes/client/api_client.py", line 362, in request
    return self.rest_client.GET(url,

  File "/root/plow/workloads/upgrade-perf/upgrade/lib64/python3.8/site-packages/kubernetes/client/rest.py", line 237, in GET
    return self.request("GET", url,

  File "/root/plow/workloads/upgrade-perf/upgrade/lib64/python3.8/site-packages/kubernetes/client/rest.py", line 231, in request
    raise ApiException(http_resp=r)

In certain situations. Presumably the api is unreachable as its upgrading. So i need to either put in some retry/non-fail logic on it or go back to the cmnd variant

Reply:
@mohit-sheth as we discussed in slack this actually isnt from this switch from the cmnd call. This is farther from the change. It looks like etcd got re-elected and the api was unavailable. I've added in a for loop on the call to "help" with this